src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -25,11 +25,264 @@
+			Assert(list_length(deadlockPath) > 1);
+
+			/*
+			 * We search for the youngest participant for two reasons
+			 * (i) predictable results (ii) kill the youngest transaction
+			 * (i.e., if a DDL continues for 1 hour and deadlocks with a
+			 * SELECT continues for 10 msec, we prefer to kill the SELECT).
+			 */
+			foreach(participantTransactionCell, deadlockPath)
+			{
+				TransactionNode *currentNode =
+					(TransactionNode *) lfirst(participantTransactionCell);
+
+				AssocateDistributedTransactionWithBackendProc(currentNode);
+
+				if (youngestTransaction->transactionId.timestamp >
+					currentNode->transactionId.timestamp)
+				{
+					youngestTransaction = currentNode;
+				}
+			}
+
+			/* we should find the backend */
+			Assert(youngestTransaction->initiatorProc != NULL);
+
+			KillBackendDueToDeadlock(youngestTransaction->initiatorProc);
+
+			hash_seq_term(&status);
+
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * CheckDeadlockForTransactionNode traverses the graph by a DFS,
+ * starting with the given node. The function checks for a cycle
+ * (i.e., the given node can be reached again while traversing the graph).
+ *
+ * Finding a cycle indicates a distributed deadlock and the function returns
+ * true on that case. Also, once a deadlock found, all the participant
+ * transactions are provided to the caller via the deadlockPath list.
+ */
+static
+bool
+CheckDeadlockForTransactionNode(TransactionNode *startingTransactionNode,
+								List **deadlockPath)
+{
+	/* the starting node is in the deadlock path by default if a transaction found */
+	*deadlockPath = list_make1(startingTransactionNode);
+
+	return CheckDeadlockForTransactionNodeRecurse(startingTransactionNode,
+												  startingTransactionNode->waitsFor,
+												  deadlockPath);
+}
+
+
+/*
+ * CheckDeadlockForTransactionNodeRecurse is the workhorse for doing the DFS.
+ * The function recursively walks over the waiting transaction nodes and
+ * keeps track of the transactions that are part of the deadlock by updating
+ * the deadlockPath list. The deadLockPath is only valid if a deadlock found.
+ */
+static
+bool
+CheckDeadlockForTransactionNodeRecurse(TransactionNode *startingTransactionNode,
+									   List *waitingTransactionNodes,
+									   List **deadlockPath)
+{
+	while (waitingTransactionNodes != NIL)
+	{
+		TransactionNode *waitingTransactionNode =
+			(TransactionNode *) linitial(waitingTransactionNodes);
+
+		/* add the transaction which is being processed to the deadlock path */
+		*deadlockPath = lcons(waitingTransactionNode, *deadlockPath);
+
+		if (waitingTransactionNode == startingTransactionNode)
+		{
+			return true;
+		}
+
+		/* don't need to revisit the node again */
+		if (waitingTransactionNode->transactionVisited)
+		{
+			continue;
+		}
+
+		waitingTransactionNode->transactionVisited = true;
+
+		if (CheckDeadlockForTransactionNodeRecurse(startingTransactionNode,
```

Comment:
Hmm, that's right. My intention was to keep track `deadlockPath `, which seemed simpler with this method.